src/main/java/io/github/moulberry/notenoughupdates/profileviewer/SacksPage.kt

```diff
@@ -0,0 +1,505 @@
+            }
+        }
+
+        GlStateManager.color(1f, 1f, 1f, 1f)
+        MC.textureManager.bindTexture(GuiProfileViewer.resource_packs)
+
+        if (page > 0) {
+            Utils.drawTexturedRect(
+                (guiLeft + 290 - 20).toFloat(),
+                (guiTop + arrowsHeight).toFloat(),
+                12f,
+                16f,
+                29 / 256f,
+                53 / 256f,
+                0f,
+                32 / 256f,
+                GL11.GL_NEAREST
+            )
+        }
+        if (page < maxPage) {
+            Utils.drawTexturedRect(
+                (guiLeft + 290 + 8).toFloat(),
+                (guiTop + arrowsHeight).toFloat(),
+                12f,
+                16f,
+                5 / 256f,
+                29 / 256f,
+                0f,
+                32 / 256f,
+                GL11.GL_NEAREST
+            )
+        }
+
+
+        if (tooltipToDisplay.isNotEmpty()) {
+            tooltipToDisplay = tooltipToDisplay.map { "§7$it" }
+            Utils.drawHoveringText(tooltipToDisplay, mouseX, mouseY, instance.width, instance.height, -1)
+            tooltipToDisplay = listOf()
+        }
+    }
+
+    fun mouseClick(mouseX: Int, mouseY: Int, mouseButton: Int): Boolean {
+        super.mouseClicked(mouseX, mouseY, mouseButton)
+
+        if (sacksJson == null || sackContents.isEmpty()) {
+            return false
+        }
+        // after this point everything in the constants json exists and does not need to be checked again
+
+        if (currentSack == "All") {
+            val sackTypes = sacksJson.getAsJsonObject("sacks")
+            var xIndex = 0
+            var yIndex = 0
+
+            val startIndex = page * pageSize
+            val endIndex = (page + 1) * pageSize
+            var itemCount = -1
+
+            for ((sackName, _) in sackTypes.entrySet()) {
+                itemCount++
+                if (itemCount < startIndex || itemCount >= endIndex) continue
+
+                if (yIndex == rows) continue
+                val x = guiLeft + 168 + xIndex * 37
+                val y = guiTop + 20 + yIndex * 41
+
+                if (mouseX > x && mouseX < x + 20) {
+                    if (mouseY > y && mouseY < y + 20) {
+                        currentSack = sackName
+                        Utils.playPressSound()
+                        page = 0
+                        maxPage = getPages(currentSack, sackTypes)
+                        return true
+                    }
+                }
+
+                xIndex++
+                if (xIndex == columns) {
+                    xIndex = 0
+                    yIndex++
+                }
+            }
+        }
+
+        if (Mouse.isButtonDown(0)) {
+            if (mouseY > guiTop + arrowsHeight && mouseY < guiTop + arrowsHeight + 16) {
+                if (mouseX > guiLeft + 290 - 20 && mouseX < guiLeft + 290 + 20) {
+                    if (mouseX > guiLeft + 290) {
```

Comment:
This entire PR is filled with opaque magic numbers, which make it extraordinarily hard to verify the lineups of click triggers and rendered objects. It also just makes things confusing. Why is there an arrowHeight, but not width, for example? That just makes me guess that the constant that actually did represent the arrow width could not be it, because why would one of them have a name and the other not. Please try to avoid magic constants, you can turn a detekt lint on for them in config/detekt/detekt.yml, and just check over this file after running the ./gradlew detekt task.